### DISCUSSION TITLE
Labgrid for testing userspace software

### DISCUSSION THREAD

**Discussion by phlmm** [labgrid-project/labgrid#1203] [Q&A] [answered]
Hello, 

I am considering labgrid as a my default framework for embedded linux testing. I've ran the demos successfully and am very happy with the way labgrid works. 

However I'm having trouble to comprehend how to integrate it into a little bit more software testing usecases. 
The labgrid setup I use is laptop -> Labhost -> DUT.

Lets say I have a custom application on an UART link using custom protocol used to command and control the DUT from the Labhost.

What is the correct way to setup tests for this scenario using labgrid?

**Comment by Bastian-Krause** [comment, marked as answer]
Do you have an example for "command and control" in your use case? What do you want to achieve/test exactly? Run an application on the "labhost" as a counterpart to an application started on your DUT? Or craft a custom counterpart using some Python library, so you can craft requests/responses dynamically?

Depending on your use case, you might want to use labgrid to acquire the resources, turn on power, bring the device to a desired state in userspace. Everything else could probably be done with Python/pytest and maybe some other libraries, without further labgrid interaction?

**Comment by phlmm** [comment]
Initially, I thought I had to create custom resources/drivers for my device. 

In the end I ended up using Labgrid for infrastructure management and pytest for the rest.

An external script manages the actual tests execution using pytest. This includes results gathering.

The combination of usb-sd-mux and Labgrid is extremely powerful, allowing me to create very nice CI combos as well. (i.e. flashing the device directly from my build machine).

**Reply by Bastian-Krause** [reply]
Glad to hear it works well for you!
